contracts/generator_controller/src/contract.rs

```diff
@@ -136,6 +141,82 @@ pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> E
     }
 }
 
+/// ## Description
+/// This function removes the votes of owners that are blacklisted. Returns [`Response`] in case
+/// of success or [`ContractError`] in case of errors.
+///
+/// ## Params
+/// * **deps** is an object of type [`DepsMut`].
+///
+/// * **env** is an object of type [`Env`].
+///
+/// * **holders** is a vector of type [`String`]. Contains blacklisted holders whose votes will be
+/// removed.
+fn kick_holders(deps: DepsMut, env: Env, holders: Vec<String>) -> ExecuteResult {
+    let block_period = get_period(env.block.time.seconds())?;
+    let escrow_addr = CONFIG.load(deps.storage)?.escrow_addr;
+    let blacklisted_holders = get_blacklisted_holders(deps.querier, &escrow_addr)?;
+
+    // check if holders are blacklisted
+    let mut holder_addrs: Vec<Addr> = vec![];
+    for holder in holders {
+        let holder_addr = addr_validate_to_lower(deps.api, &holder)?;
+        if !blacklisted_holders.contains(&holder_addr) {
+            return Err(ContractError::HolderIsNotBlacklisted {});
+        }
+        holder_addrs.push(holder_addr);
+    }
+
+    for holder_addr in holder_addrs {
+        if USER_INFO.has(deps.storage, &holder_addr) {
+            let user_info = USER_INFO.load(deps.storage, &holder_addr)?;
+
+            if user_info.lock_end > block_period {
+                let user_last_vote_period = get_period(user_info.vote_ts).unwrap_or(block_period);
+                // Calculate voting power before changes
+                let old_vp_at_period = calc_voting_power(
+                    user_info.slope,
+                    user_info.voting_power,
+                    user_last_vote_period,
+                    block_period,
+                );
+
+                // Cancel changes applied by previous votes
+                user_info.votes.iter().try_for_each(|(pool_addr, bps)| {
+                    cancel_user_changes(
+                        deps.storage,
+                        block_period + 1,
+                        pool_addr,
+                        *bps,
+                        old_vp_at_period,
+                        user_info.slope,
+                        user_info.lock_end,
+                    )
+                })?;
+
+                let ve_lock_info = get_lock_info(deps.querier, &escrow_addr, &holder_addr)?;
+                let user_vp = get_voting_power(deps.querier, &escrow_addr, &holder_addr)?;
+
+                let user_info = UserInfo {
+                    vote_ts: env.block.time.seconds(),
+                    voting_power: user_vp,
```

Comment:
Both voting power and slope are zero if a user is blacklisted. We do not need to query voting escrow. Just set zero here